src/cargo/sources/registry/http_remote.rs

```diff
@@ -344,40 +425,57 @@ impl<'cfg> RegistryData for HttpRegistry<'cfg> {
                     // The crate was not found or deleted from the registry.
                     return Poll::Ready(Ok(LoadResponse::NotFound));
                 }
+                401 if !self.auth_required && path == Path::new("config.json") => {
+                    debug!("re-attempting request for config.json with authorization included.");
+                    self.fresh.remove(path);
+                    self.auth_required = true;
+
+                    // Look for www-authenticate header of the form: "Cargo login_url=..."
+                    self.login_url = (|| {
+                        www_authenticate::parse(result.header_map.get(WWW_AUTHENTICATE)?)
+                            .get("Cargo")?
+                            .get("login_url")?
+                            .into_url()
+                            .ok()
+                    })();
+                }
+                401 if self.auth_required => {
+                    let body = String::from_utf8_lossy(&result.data);
+                    let body = truncate_with_ellipsis(&body, 1000);
+                    let err = anyhow::anyhow!("remote server said: {}", body,).context(
+                        auth::AuthorizationError {
+                            sid: self.source_id.clone(),
+                            login_url: self.login_url.clone(),
+                            message: "token rejected",
+                        },
+                    );
+                    return Poll::Ready(Err(err));
+                }
                 code => {
-                    return Err(anyhow::anyhow!(
-                        "server returned unexpected HTTP status code {} for {}\nbody: {}",
+                    let body = String::from_utf8_lossy(&result.data);
+                    let body = truncate_with_ellipsis(&body, 1000);
+                    return Poll::Ready(Err(anyhow::anyhow!(
+                        "remote server responded with unexpected HTTP {} for '{}'\nmessage from server: {}",
                         code,
                         self.full_url(path),
-                        str::from_utf8(&result.data).unwrap_or("<invalid utf8>"),
-                    ))
-                    .into();
+                        body,
+                    )));
                 }
             }
-
-            return Poll::Ready(Ok(LoadResponse::Data {
-                raw_data: result.data,
-                index_version: Some(result.index_version),
-            }));
-        }
-
-        if self.config.offline() {
-            return Poll::Ready(Err(anyhow::anyhow!(
```

Comment:
This check was unnecessary, since it's already done by `ops::http_handle` below.